src/generator.cs

```diff
@@ -1401,19 +1546,25 @@ public string MarshalParameter (MethodInfo mi, ParameterInfo pi, bool null_allow
 		throw new BindingException (1002, true, "Unknown kind {0} in method '{1}.{2}'", pi, mi.DeclaringType.FullName, mi.Name.GetSafeParamName ());
 	}
 
-	public bool ParameterNeedsNullCheck (ParameterInfo pi, MethodInfo mi)
+	public bool ParameterNeedsNullCheck (ParameterInfo pi, MethodInfo mi, PropertyInfo propInfo = null)
 	{
 		if (pi.ParameterType.IsByRef)
 			return false;
 
 		if (HasAttribute (pi, typeof (NullAllowedAttribute)))
 			return false;
 
-		if (mi.IsSpecialName && mi.Name.StartsWith ("set_", StringComparison.Ordinal)){
+		if (IsSetter (mi)) {
 			if (HasAttribute (mi, typeof (NullAllowedAttribute))){
 				return false;
 			}
 		}
+
+		if (HasBindAsAttribute (pi) || (propInfo != null && HasBindAsAttribute (propInfo))) {
```

Comment:
You're first checking if pi/propInfo has the BindAs attribute, then you're actually getting the attribute, but the checking requires getting the attribute as well, so this is wasteful.

Instead just get the attribute, and check if you got it:

    var bindAsAtt = GetBindAsAttribute (pi) ?? GetBindAsAttribute (propInfo);
    if (bindAsAtt != null)
        return bindAsAtt.IsNullable;